test/pkg/environment/clustermonitor.go

```diff
@@ -0,0 +1,164 @@
+package environment
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	v1 "k8s.io/api/core/v1"
+	"knative.dev/pkg/logging"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// ClusterMonitor is used to monitor the cluster state during a running test
+type ClusterMonitor struct {
+	ctx        context.Context
+	kubeClient client.Client
+
+	mu                 sync.RWMutex
+	recordings         []recording
+	nodesSeen          sets.String
+	numberNodesAtReset int
```

Comment:
Perhaps we can simply filter nodes with the label karpenter.sh/provisioner-name and only analyze those?

Reply:
I think that's acceptable given the amount of code and how specific our analysis is. Are you imagining a higher calling?